inflight.go

```diff
@@ -195,6 +186,16 @@ func callAndCacheBatch[V, T any](ctx context.Context, c *Client[T], opts callBat
 			}
 			return response, ErrInvalidType
 		}
+
+		// It could be only cached records here, if we we're able
+		// to get some of the IDs from the distributed storage.
+		if call.err != nil && !errors.Is(call.err, ErrOnlyCachedRecords) {
+			return response, call.err
+		}
+
+		if errors.Is(call.err, ErrOnlyCachedRecords) {
+			err = ErrOnlyCachedRecords
+		}
 	}
 
 	return response, err
```

Comment:
Same here

> [!Note]
> I know you are moving code, so the code is not yours, but as you are refactoring it should be reviewed and improved

There is something strange.

```go
		if call.err != nil && !errors.Is(call.err, ErrOnlyCachedRecords) {
			return response, call.err
		}

		if errors.Is(call.err, ErrOnlyCachedRecords) {
			err = ErrOnlyCachedRecords
		}
 	}
 
 	return response, err
```

Let's try to check all cases.

if call.err is nil, the first and second if are ignored, so we return err

Then, if call.err is ErrOnlyCachedRecords, the first if is ignored, the second sets err to ErrOnlyCachedRecords, but here we could have returned call.err

Last case, if call.err is not nil, but not ErrOnlyCachedRecords, the call.err is returned

So call.err is returned when ever present

But then, here this code could be simplified to this, no?

```suggestion
		// It could be only cached records here, if we we're able
		// to get some of the IDs from the distributed storage.
		if call.err != nil {
			return response, call.err
		}
 	}
 
 	return response, err
```

Reply:
It might be simpler but it's not functionally equivalent. The original code intentionally handles errors of type `ErrOnlyCachedRecords` differently from other errors. Specifically, it does the following:

- **For non‑ErrOnlyCachedRecords errors:** It immediately returns the error, short‑circuits the loop, and doesn’t process further calls.
- **For ErrOnlyCachedRecords errors:** It doesn’t return immediately; instead, it records that error (by setting err = ErrOnlyCachedRecords) and continues processing the rest of the batch. This way, even if one call returns ErrOnlyCachedRecords, the loop still gathers IDs from other calls. If later another call yields a non‑ErrOnlyCachedRecords error, that error would take precedence and be returned immediately.

The original logic will only short‑circuit the function when a “real” error occurs, while still accumulating partial responses when the only issue is cached‑only records.

However, one could argue that the block could be simplified like this:

```go
if call.err != nil {
	if !errors.Is(call.err, ErrOnlyCachedRecords) {
		return response, call.err
	}
	err = ErrOnlyCachedRecords
}
```

Reply:
> The original logic will only short‑circuit the function when a “real” error occurs, while still accumulating partial responses when the only issue is cached‑only records.

Another question: Why do we want to short-circuit in any case? If call A had e.g. a 500 response error, why not still process call B?

To answer my own question, I think it could get complicated:

- Call A: Error X
- Call B: Success
- Call C: Error Y

In this scenario, how should `callAndCacheBatch` respond, with Error X or Error Y?

So I think it makes sense to "fail fast" like you do here.

Reply:
Hi! Yeah, there are multiple places like this where choosing the "right thing" to do is quite complex, and I don’t want to flood the API with custom errors that clients would then have to handle with large switch cases. My approach has been to keep it as simple as possible until someone opens an issue and explicitly requests more flexibility.

However, since we’ve discussed this code quite a bit, I went ahead and simplified it, as well as added some comments:
 
 ```go
 	for call, callIDs := range callIDs {
		call.Wait()

		// We need to iterate through the values that WE want from this call. The batch
		// could contain hundreds of IDs, but we might only want a few of them.
		for _, id := range callIDs {
			if v, ok := call.val[id]; ok {
				response[id] = v
			}
		}

		// This handles the scenario where we either don't get an error, or are
		// using the distributed storage option and are able to get some records
		// while the request to the underlying data source fails. In the latter
		// case, we'll continue to accumulate partial responses as long as the only
		// issue is cached-only records.
		if err == nil || errors.Is(call.err, ErrOnlyCachedRecords) {
			err = call.err
			continue
		}

		// For any other kind of error, we'll short‑circuit the function and return.
		return response, call.err
	}
 ```

Reply:
This version is way cleaner 👍